controller/src/main/java/io/pravega/controller/store/stream/InMemoryStream.java

```diff
@@ -99,6 +101,11 @@
     @GuardedBy("txnsLock")
     private final Map<Integer, Set<UUID>> epochTxnMap = new HashMap<>();
 
+    private final Object writersLock = new Object();
+
+    @GuardedBy("writersLock")
```

Comment:
I have followed that same pattern across the class. Basically in case of failure scenarios like Data version mismatch etc, handling it inside computeIfAbsent/Present makes the code more complex. 
This is more readable.